lib/tokenize.cpp

```diff
@@ -1746,34 +1746,86 @@ void Tokenizer::simplifyTypedef()
 
 namespace {
     struct ScopeInfo3 {
-        ScopeInfo3(const std::string &name_, const Token *bodyStart_, const Token *bodyEnd_)
-            : name(name_), bodyStart(bodyStart_), bodyEnd(bodyEnd_) {}
+        enum Type { Global, Namespace, Record, MemberFunction, Other };
+        ScopeInfo3() : parent(nullptr), type(Global), bodyStart(nullptr), bodyEnd(nullptr) {}
+        ScopeInfo3(ScopeInfo3 *parent_, Type type_, const std::string &name_, const Token *bodyStart_, const Token *bodyEnd_)
+            : parent(parent_), type(type_), name(name_), bodyStart(bodyStart_), bodyEnd(bodyEnd_) {
+            fullName = name;
+            ScopeInfo3 *scope = parent;
+            while (!fullName.empty() && scope && scope->parent) {
```

Comment:
why do we have this `!fullName.empty()` check? Because we don't want to run this loop when the `name` is empty? When does that happen.. when we are in an unnamed scope?
Probably your code is fine I just want to understand it..

Reply:
Anything that is not a type or namespace will not have a name.  That includes if, for do, while, {} and functions.  We only care about names in qualifications so we can look them up.

Reply:
Actually I don't think control flow scopes are even looked for so only global functions don't have names.  Member function scopes have the name of the record type they belong to.  We should probably look for all scopes someday because they could contain a `using namespace` but nobody has complained yet or created a ticket for it.

Reply:
ok. It can be debated if the `!fullName.empty()` should be broken out from the loop (before) but that's not very important.

Reply:
I see your point now.  I can fix that later.  There will be a lot of follow on PRs if we want `using namespace` to work correctly 100% of the time.

The long term plan is to get simplifyUsing complete and then add typedef support to it because it also has the same bugs but the fixes can't be ported to it because it has a totally different way of handling scopes.